charmcraft/commands/store/__init__.py

```diff
@@ -247,75 +248,66 @@ def run(self, parsed_args):
             logger.info(line)
 
 
+def get_name_from_zip(filepath):
+    """Get the charm/bundle name from a zip file."""
+    try:
+        zf = zipfile.ZipFile(filepath)
+    except zipfile.BadZipFile:
+        raise CommandError("Cannot open {!r} (bad zip file).".format(str(filepath)))
+
+    # get the name from the given file (trying first if it's a charm, then a bundle,
+    # otherwise it's an error)
+    if 'metadata.yaml' in zf.namelist():
+        try:
+            name = yaml.safe_load(zf.read('metadata.yaml'))['name']
+        except Exception:
+            raise CommandError(
+                "Bad 'metadata.yaml' file inside charm zip {!r}: must be a valid YAML with "
+                "a 'name' key.".format(str(filepath)))
+    elif 'bundle.yaml' in zf.namelist():
+        try:
+            name = yaml.safe_load(zf.read('bundle.yaml'))['name']
+        except Exception:
+            raise CommandError(
+                "Bad 'bundle.yaml' file inside bundle zip {!r}: must be a valid YAML with "
+                "a 'name' key.".format(str(filepath)))
+    else:
+        raise CommandError(
+            "The indicated zip file {!r} is not a charm ('metadata.yaml' not found) "
+            "nor a bundle ('bundle.yaml' not found).".format(str(filepath)))
+
+    return name
+
+
 class UploadCommand(BaseCommand):
-    """Upload a charm to Charmhub."""
+    """Upload a charm or bundle to Charmhub."""
 
     name = 'upload'
-    help_msg = "Upload a charm to Charmhub"
+    help_msg = "Upload a charm or bundle to Charmhub"
     overview = textwrap.dedent("""
-        Upload a charm to Charmhub.
+        Upload a charm or bundle to Charmhub.
 
-        Push a charm to Charmhub where it will be verified for conformance
-        to the packaging standard. This command will finish successfully
-        once the package is approved by Charmhub.
+        Push a charm or bundle to Charmhub where it will be verified for
+        conformance to the packaging standard. This command will finish
+        successfully once the package is approved by Charmhub.
```

Comment:
```suggestion
        Push a charm or bundle to Charmhub where it will be verified.
        This command will finish successfully once the package is
        approved by Charmhub.
```

Reply:
improved